### ISSUE TITLE
WebUI refactoring (POC)

### ISSUE THREAD

**Issue by oshtaier** [SatelliteQE/robottelo#5791]
After some investigation in the area, two main approaches for review were selected:

**widgetastic+navmazing**

*Scope*
In case we like to use both frameworks as they were designed - we need to have complete transition to OOP paradigms and that mean we need to re-write 95% of existing code base for UI.

*Pros*
- Level of abstractions let us have proper isolation for framework and test layers, so it is easier to fix corresponding code parts depends on changes in application (whether application logic or internal implementation were changed)

*Cons*
- Obviously, we have to take into consideration other projects(at least CF) in case we like to change basic framework elements.
- Will require a lot of time to have stable runs as framework code was never tested against any other application. I can't find a way to execute, for example, one easiest Architecture test from Satellite using new framework without introducing static delays.
- Code will be almost impossible to read and find logical connection between elements, especially for new people who like to contribute anything new.
- Code will be hard to debug as you need to go through hundreds abstraction layers to get to things you are looking for
- Testing flow will not be flexible as you are going to be very strict to your object model

*Summary*
Approach is not recommended for projects like ours as introduce more complexity rather than benefits

*References*
https://github.com/ManageIQ/integration_tests/blob/2bc9354de1be204b53a47eb249d5c1ef892e7fb5/cfme/utils/appliance/implementations/ui.py#L199
https://github.com/ManageIQ/integration_tests/blob/master/widgetastic_manageiq/__init__.py
...

**Existing code refactoring**

Description:
- Current navigation re-implementation to use simple tree(xml, json or dict) to go through the nods without extra logic
- Entity logic isolation (move code from tests to the middle layer, move and making locators as class properties)
- Make `Base` module as isolated framework class with only necessary methods

*Scope*
Framework and middle layers will have average impact. Test level should have small impact. Not more than 35% of code changes are expected

*Pros*
- Easy to read, easy to write, easy to debug
- We and only we have full control on things that we do

*Cons*
- It is necessary to spend time on code implementation that is already present in third party libraries, so such efforts can be treated as waste of time

*Summary*
Moving to concept used, but not to framework themselves sounds as better idea

**Comment by abalakh** [comment]
+1 to what @oshaier said.

tl;dr version - comparing our approach vs pageobject lib is like comparing python vs java. 

Using strict OOP will result in solving the same problems in more excessive way - even for basic tests it means multiple layers of abstraction of test framework, much more code which is less readable and harder to understand, debugging issues will become a nightmare as you'll have keep in mind entire tree of inheritance, go through tens of classes to finally find that one method causing issues.
Not to mention we'll have to write everything from scratch, and for what purpose? Tests won't become faster or less resource-consuming, it only will get harder to debug and fix them and i doubt tests will become more stable, at least for the first 1-1.5 years when the framework is fresh and not polished enough.
Contributing to such framework will definitely become way harder, especially for newcomers - keeping in mind inheritance tree and flowcharts with methods order (like [this one for navmazing](https://github.com/psav/navmazing/raw/master/flowchart.png)) is a must.

I'm not saying our framework is the best and in perfect condition, it has a room for improvement, but why solve things the hard way? `Simple is better than complex`, as Zen of Python says, isn't it? :)

I believe we'd benefit way more from refactoring our framework + switching to pytest + py3 + latest selenium than building yet another typical pageobject framework from scratch, i'd strongly recommend not to withdraw robottelo-UI in favor of widgetastic+navmazing.

**Comment by ldjebran** [comment]
I do want to concentrate on why we do not want to switch as many things has been said in this regard, but want to concentrate on the positive things that can bring such a switch not only for our project but also to other projects using patternfly (https://www.patternfly.org).
1- Shared UI testing knowledge base between different projects teams.
2- Common standard for all UI testing.
3- New projects can start UI testing very quickly, as many of the issues has been resolved

This is only to name the few pluses.

Switching to a common framework, we will bring also our experience in other fields that has not been yet resolved (like saucelabs implementation and session problems), after our resolution other teams can switch very quickly.

I'll recommend to view this issue globally and not only in the frame of our project.

For example Cloudforms team has resolved many problems (that we can use directly) but still using containers for UI, we can implement the saucelabs resolution and it will be a plus for them. Not to count other teams that can resolve other problems that we will have to face in the future.

For the moment I do not know all the teams using "widgetastic" but I am convinced that it will bring benefit in the future and will be a good experience for all of us while interacting with other teams trying to resolve common problems.

**Comment by omaciel** [comment]
[Keeping it short due to being jet lagged hehe]

@oshtaier thank you for the analysis! Your points are valid and I understand your concerns.

@abalakh I also thank you for your comments and in many ways I agree with your points, specially when you applied the Zen of Python to illustrate your opinion :) The short-term solution would be to invest our time in "patching" or refactoring Robottelo UI (I like the term actually), but I am thinking long term value... and there is also the new "angle" where our UI tests would be much smaller (and hopefully 100% reliable). By the way, I don't think that we need to completely drop what we have right away... I am looking for a gradual transition here.

@ldjebran thank you for your comments and I too believe that we may actually be in a position to develop a more robust framework. The key ingredient, and I will try to speak to Mulan while I'm in Brno, is to get a nice Proof of Concept that can show how the CF team was able to use these frameworks to solve really complicated problems for the project.

Since I created Robottelo from scratch, I do suffer from that feeling that software engineers have whenever something comes up that "threatens the life" of their project :) I don't say this lightly, but I do think that we need to consider Robottelo 2.0 our next objective as a team and even some integration with Rails/rSpec.

More to follow.

**Comment by ntkathole** [comment]
Thanks for nice discussion. Closing this as we have widgetastic+navmazing solution implemented now : )
